lib/task.ml

```diff
@@ -1,33 +1,108 @@
-type 'a task = unit -> 'a
+open EffectHandlers
+open EffectHandlers.Deep
 
-type 'a promise = ('a, exn) result option Atomic.t
+type 'a task = unit -> 'a
 
-exception TasksActive
+type message =
+  Work of (unit -> unit)
+| Quit
 
-type task_msg =
-  Task : 'a task * 'a promise -> task_msg
-| Quit : task_msg
+type task_chan = message Multi_channel.t
 
 type pool_data = {
   domains : unit Domain.t array;
-  task_chan : task_msg Multi_channel.t;
+  task_chan : task_chan;
   name: string option
 }
 
 type pool = pool_data option Atomic.t
 
+type 'a promise_state =
+  Returned of 'a
+| Raised of exn * Printexc.raw_backtrace
+| Pending of ((unit, unit) continuation * task_chan) list
+
+type 'a promise = 'a promise_state Atomic.t
+
+type _ eff += Wait : 'a promise * task_chan -> unit eff
+
+let get_pool_data p =
+  match Atomic.get p with
+  | None -> raise (Invalid_argument "pool already torn down")
+  | Some p -> p
+
+let cont (k, c) = Multi_channel.send c (Work (continue k))
+let discont e bt (k, c) = Multi_channel.send c (Work (fun _ ->
+  discontinue_with_backtrace k e bt))
+
 let do_task f p =
-  try
-    let res = f () in
-    Atomic.set p (Some (Ok res))
-  with e ->
-    Atomic.set p (Some (Error e));
-    match e with
-    | TasksActive -> raise e
-    | _ -> ()
+  let action, result =
+    try cont, Returned (f ())
+    with e ->
+      let bt = Printexc.get_raw_backtrace () in
+      discont e bt, Raised (e, bt)
+  in
+  match Atomic.exchange p result with
+  | Pending l -> List.iter action l
+  |  _ -> failwith "Task.do_task: impossible, can only set result of task once"
 
-let named_pools = Hashtbl.create 8
+let async pool f =
+  let pd = get_pool_data pool in
+  let p = Atomic.make (Pending []) in
+  Multi_channel.send pd.task_chan (Work (fun _ -> do_task f p));
+  p
 
+let rec await pool promise =
+  let pd = get_pool_data pool in
+  match Atomic.get promise with
+  | Returned v -> v
+  | Raised (e, bt) -> Printexc.raise_with_backtrace e bt
+  | Pending _ ->
+      perform (Wait (promise, pd.task_chan));
```

Comment:
Shouldn't the wait here already return the `'a` value that we care about? (if the types above are changed accordingly).
Would be one less atomic operation on wakeup (may not matter in practice though), but it'd avoid the impression that this can lead to an infinite loop due to bugs: it can't, you either get the result immediately (an optimization), or you perform the effect which blocks as long as necessary and then gives you back the final result, without spurious wakeups.

Reply:
Fixed in https://github.com/ocaml-multicore/domainslib/pull/51/commits/db22e176f720d5097c0b1d3b8edf01f4248dcad6.